BIND query log decoder Lua plugin

nickchappell opened the pull request:
@rafrombrc @trink do you want this squashed into a single commit?

nickchappell committed:
Add a Lua script that can decode BIND DNS query logs along with
accompanying documentation.

sathieu commented:
Hello @nickchappell ,

The parser should probably go to lua_sandbox, to be available to other tools.

trink commented on sandbox/lua/decoders/bind_query_log.lua:
```diff
@@ -0,0 +1,274 @@
+
+  return 0
+
+end
```
add new line

nickchappell replied:
Done.

trink commented on sandbox/lua/decoders/bind_query_log.lua:
```diff
@@ -0,0 +1,274 @@
+  --The LogstreamerInput that generates the messages automatically puts the log line in the Payload:
+  local query_log_line = read_message("Payload")
+  --Use the bind_query grammar defined farther above to create a local grammar object:
+  local grammar = l.Ct(bind_query)
```
Ideally this grammar would be made into a module so it could be re-used.  To  be used with Heka it would need to go into the heka branch of the lua_sandbox

nickchappell replied:
Done.

trink commented on sandbox/lua/decoders/bind_query_log.lua:
```diff
@@ -0,0 +1,274 @@
+  --This value is read in from the 'msg_type' config option in the TOML:
+  Type        = msg_type,
+  Payload     = nil,
+  Severity    = 'info',
```
Severity should be numeric

nickchappell replied:
Is this moot if the script has been turned into a module?

trink commented on sandbox/lua/decoders/bind_query_log.lua:
```diff
@@ -0,0 +1,274 @@
+library and combines them with the hyphen (-) character to match anything that's
+part of a valid hostname label. The ^1 means match one or more instances of it.
+--]]
+local hostname_fragment = (l.upper + l.lower + l.digit +  "-")^1
```
alnum would suffice for the first three but a common hostname grammar should be used instead (there is an open issue since we are getting quite the collection of hostname parsers) https://github.com/mozilla-services/lua_sandbox/issues/129

nickchappell replied:
Once the other things are cleaned up, could we merge this as-is with my custom hostname grammar, then modify it to use the generic hostname grammar after https://github.com/mozilla-services/lua_sandbox/issues/129 is merged?

nickchappell replied:
Also, I started down the path of breaking the hostname into fragments so I could get the name and query domain out of the log lines as separate values (for us, having the domain be a separate searchable field is valuable).

trink replied:
that is fine the refactor can happen in issue 129

trink commented on sandbox/lua/decoders/bind_query_log.lua:
```diff
@@ -0,0 +1,274 @@
+
+Source: https://en.wikipedia.org/wiki/List_of_DNS_record_types
+--]]
+dns_record_type = l.Cg(
```
A regular capture will suffice for all RecordType and RecordClass entries e.g. `l.C"A"`

nickchappell replied:
How would that work with the multiple record types? A `l.C"A"` and so on for each record type?

trink replied:
Yeah just change all the l.P's to l.C's and get rid of the substitution expression

nickchappell commented:
@trink should I close this PR and open one up against https://github.com/mozilla-services/lua_sandbox/ ?

trink commented:
The PR against heka is actually fine for the moment it will be easier to migrate everything from one place.

nickchappell commented:
@trink should the refactor to make the grammar a module wait until then as well, or should I do that now? If it should be modular-ized now, are there any examples of existing modules I should follow?

trink commented:
It should be modularized now.  Here is an example: https://github.com/mozilla-services/lua_sandbox/blob/master/modules/ip_address.lua

nickchappell commented:
Would you have to write another Lua script that calls the modular-ized BIND decoder in order to actually use the grammar?

nickchappell commented:
@trink I addressed the issues you mentioned in the inline comments, minus the hostname grammar refactor, as you mentioned that would be taken care of in #129. Do you want me to squash all of the commits?

trink commented:
Yes please squash them

nickchappell commented:
@trink @rafrombrc Commits have been squashed.

nickchappell commented:
@trink anything else needed so this can be merged?

nickchappell commented:
@trink any updates?

nickchappell commented:
@trink @rafrombrc Any update on getting this merged?

trink commented:
@nickchappell Sorry for the delay, this request is in a state of limbo as it will need to be refactored and moved to the lua_sandbox in the very near future.  @rafrombrc thoughts? 
- merge now and migrate later?
- turn it into a grammar module for the new sandbox and have it available when the integration with Heka is complete?
- ??

nickchappell commented:
Hi @trink,

It should already be modular-ized. I think you mentioned earlier that this could be merged into Heka now and then added to the new sandbox at some point later.

nickchappell commented:
Thanks!

trink commented:
The module has to be pulled out and placed in the modules directory and the decoder has to require it and implement the expected sandbox API.  The question above was for Rob to determine timeline/frame and where the work should be committed (I guess he is fine with Heka, so it should be fixed there).

rafrombrc commented:
Yes, sorry, I misinterpreted the thread as an r+ from Trink, so I merged. Happy to have it land in Heka for now.
